### ISSUE TITLE
[bug] pylint no-member warning for `tool_requires` and `test_requires`

### ISSUE THREAD

**Issue by hofbi** [conan-io/conan#12004]
### Environment Details (include every applicable attribute)
  * Operating System+version: Ubuntu 22.04
  * Compiler+version: gcc-11.2
  * Conan version: 1.52.0
  * Python version: 3.10.4

### Steps to reproduce (Include if Applicable)

Within the `conanfile.py` I use:

```python
def build_requirements(self):
        self.test_requires("gtest/1.12.1")
```

Running pylint with the conan plugin results in:

```shell
E1101: Instance of '...' has no 'test_requires' member (no-member)
```

As quick fix, I suppress the warning using `# pylint: disable=no-member`

References:

* pylint no-member warning: https://pylint.pycqa.org/en/latest/user_guide/messages/error/no-member.html

**Comment by agagniere** [comment]
## TL;DR:

If you think about it, Pylint should not be able to know that `test_requires` is a member of your class, because this member should only exist is `build_requirements` and not other methods.

## Long answer :
Just some ideas about this:

Pylint is a static code analysis tool, it does not execute the code to validate it, only reads it.
So it only knows about the members of an object set during construction (by reading `__init__`), because those members are guaranteed to be there in the `self` passed to all other methods.

On the other hand, the `test_requires` method should only exist in certain methods, you don't want it to exist in `build` or `package` for example. So it is not guaranteed to be there at all time, it is just that conan developpers made sure to set it before calling your `build_requirements`.

So I wouldn't expect `pylint` or any static analysis tool to be able to tell me that `test_requires` is a member, because if you manually call `build_requirements` from `package` for instance, it will indeed be missing and fail (so in that sense, pylint is right).
It is just that we are inside conan's conventions, that go beyond python rules : calling `build_requirements` from `package` is possible in python, but is forbidden in conan. But of course `pylint` is not aware of such rules.

So it is not a conan bug, just a misuse of pylint.

Trying to reproduce your error with dummy code, I wrote :

```python
class Foo:
    def __init__(self):
        self.is_bar = False # is_bar is guaranteed to exist in all Foo instances

    def first(self):
        self.bar = print # bar is not guaranteed to exist
        self.is_bar = True


class Toto(Foo):

    def second(self):
        self.bar("Hello")

baz = Toto()
baz.first()
baz.second()
```
But `pylint` only told me :
```
W0201: Attribute 'bar' defined outside __init__ (attribute-defined-outside-init)
```

So looking at the `pylint` documentation you linked, we can read :
> If you are getting the dreaded no-member error, there is a possibility that either:
>   * You're launching pylint without the dependencies installed in its environment.

So I'm guessing you could "solve" this by making sure pylint has access to conan's source code when linting.

**Comment by hofbi** [comment]
Thank you for the detailed explanation. I also did not expect this a bug of conan itself but more on the conan pylint plugin. Looking at https://github.com/conan-io/conan/blob/develop/conans/pylint_plugin.py#L30, I could imagine that this could be solved by adding tool_requires and test_requires to that dict.

**Comment by memsharded** [comment]
What is clear is that there is no perfect solution to that linter. Conan recipes are a DSL on top on Python, so it doesn't necessarily follow all Python, and it cannot be linted following the same logic, and linting tools are not prepared for this use.

**Comment by agagniere** [comment]
> Thank you for the detailed explanation. I also did not expect this a bug of conan itself but more on the conan pylint plugin. Looking at https://github.com/conan-io/conan/blob/develop/conans/pylint_plugin.py#L30, I could imagine that this could be solved by adding tool_requires and test_requires to that dict.

Good point, I didn't know about this, and indeed it would seem that this plugin allows one to expose run-time logic to Pylint.
Interesting.

**Comment by hofbi** [comment]
@agagniere There is a specific section about this in the docs: https://docs.conan.io/en/latest/integrations/linting/recipe.html

**Comment by memsharded** [comment]
Conan 2 has been out for more than one year and it seems the pylint_plugin in the repo is completely outdated and broken, which means that it is not being used at all, otherwise we would have got more reports.

I am removing it from the project, Conan 2 also added a bit more of defined variables, so linters should be a bit more forgiving, but in any case and as exposed above, the problem is that linting the Conan DSL with python linters is mostly impossible.

Closing this ticket as not planned, thanks for the feedback!
